cpp/src/groupby/groupby.cu

```diff
@@ -55,6 +56,13 @@ groupby::groupby(table_view const& keys,
     _column_order{column_order},
     _null_precedence{null_precedence}
 {
+  auto is_supported_key_type = [](auto col) {
+    return col.type().id() == type_id::STRUCT     // Struct does not support ==.
+           or col.type().id() == type_id::STRING  // String does not support == on __host__.
```

Comment:
Special case logic like this is precisely what I was hoping to avoid with `is_equality_comparable()`. Furthermore, this doesn't seem right. What if the STRUCT column contains a column where `is_equality_comparable()` fails? This check should happen after the flattening.

How does `is_equality_comparable(STRING)` fail?

Reply:
> this doesn't seem right. What if the STRUCT column contains a column where is_equality_comparable() fails?

That check (for `STRUCT` containing `LIST`) exists already, as part of `flatten_nested_columns()`.
Hmm. After the last round of changes to the `flatten()`/`unflatten()` code path, I wonder if we should simply rely on `flatten()` flagging unsupported types. I'll check this in a `Release` build.

> How does `is_equality_comparable(STRING)` fail?

That's an unfortunate consequence: `string_view::operator ==` is `__device__`, while `row_equality_operator` is checking `is_equality_comparable()` from `__host__`. This causes `row_equality_operator` to bork on `STRING`.

Reply:
This has been resolved, by moving the checks to after `flatten_nested_columns()`.